crates/red_knot_python_semantic/resources/mdtest/suppressions/no-type-check.md

```diff
@@ -0,0 +1,94 @@
+# `@no_type_check`
+
+> If a type checker supports the `no_type_check` decorator for functions, it should suppress all type errors for the def statement and its body including any nested functions or classes. It should also ignore all parameter and return type annotations and treat the function as if it were unannotated.
+> [source](https://typing.readthedocs.io/en/latest/spec/directives.html#no-type-check)
+
+## Error in the function body
+
+```py
+from typing import no_type_check
+
+@no_type_check
+def test() -> int: 
+    return a + 5
+```
+
+## Error in nested function
+
+```py
+from typing import no_type_check
+
+@no_type_check
+def test() -> int:
+    def nested(): 
+        return a + 5
+```
+
+## Error in nested class
+
+```py
+from typing import no_type_check
+
+@no_type_check
+def test() -> int:
+    class Nested:
+        def inner(self):
+            return a + 5
+```
+
+## Error in decorator
+
+Both MyPy and Pyright flag the `unknown_decorator` but we don't. 
+
+```py
+from typing import no_type_check
+
+@unknown_decorator
+@no_type_check
+def test() -> int:
+    return a + 5
+```
+
+## Error in default value
+
+```py
+from typing import no_type_check
+
+@no_type_check
+def test(a: int = "test"): 
+    return x + 5
+```
+
+## Error in return value position
+
+```py
+from typing import no_type_check
+
+@no_type_check
+def test() -> Undefined: 
+    return x + 5
+```
+
+## `no_type_check` on classes isn't supported
+
+Similar to Pyright, Red Knot does not support `no_type_check` annotations on classes.
```

Comment:
sure: It's mainly that the semantics are unclear according to [this discussion](https://discuss.python.org/t/no-type-check-decorator/43119) and the typing spec explicitly leaves it open for type checkers not to support classes.

Reply:
Ah, I forgot that mypy also didn't support decorating classes with `@no_type_check` (and in fact, no other type checker does currently). Your wording here made me think that we were specifically copying pyright's implementation decisions here, but that's obviously not the case. Could you maybe reword this to something like this?

```suggestion
Red Knot does not support `no_type_check` annotations on classes currently. The behaviour of `no_type_check` when applied to classes is [not fully specified currently](https://typing.readthedocs.io/en/latest/spec/directives.html#no-type-check), and applying the decorator to classes is not supported by any other type checkers currently.
```

We could even consider emitting a diagnostic if a class is decorated with `@no_type_check`, since it clearly won't have the effect the user wants.

Reply:
> We could even consider emitting a diagnostic if a class is decorated with @no_type_check, since it clearly won't have the effect the user wants.

Possibly, but not as part of this PR ;)

Reply:
> Possibly, but not as part of this PR ;)

Is it worth adding a TODO comment somewhere as part of this PR? ;)